File: bindings/typescript/src/trust.ts
import { statusSnapshotBytes } from "./internal/lifecycle-resources.js";
import type { GrantStatusSnapshot, PrincipalStatusSnapshot } from "./lifecycle.js";
import { AuthsWorkflowError } from "./workflow/errors.js";
import { trustedContextSource, type TrustedContextSource } from "./workflow.js";
import { loadPackagedWorkflowEngine } from "./verifier/wasm.js";
import { emitAuthsEvent, type TelemetryPort } from "./observability.js";

export interface TrustProfile {
  readonly id: string;
  readonly version: number;
}

export interface TrustPermission {
  readonly capability: string;
  readonly resource: string;
}

export interface TrustBudget {
  readonly algebra: string;
  readonly value: bigint;
}

export type TrustStatusPolicy =
  | Readonly<{ mode: "expiry-only" }>
  | Readonly<{ mode: "snapshot-required"; method: string; maxAge: bigint }>;

export interface TrustAnchorConfiguration {
  readonly id: string;
  readonly principal: string;
  readonly acceptedMethods: readonly string[];
  readonly profiles: readonly TrustProfile[];
  readonly permissions: readonly TrustPermission[];
  readonly resourceNamespaces: readonly string[];
  readonly audiences: readonly string[];
  readonly notBefore: bigint;
  readonly expiresAt: bigint;
  readonly budget?: TrustBudget;
  readonly maxDelegationDepth: number;
  readonly assurancePolicy: string;
  readonly statusPolicy: TrustStatusPolicy;
}

export type AssuranceParticipantRole = "root" | "intermediate" | "actor" | "external-issuer";
export type AssuranceQuantifier = "any" | "every";

export interface AssuranceRequirementConfiguration {
  readonly role: AssuranceParticipantRole;
  readonly quantifier: AssuranceQuantifier;
  readonly claimKind: string;
  readonly maximumAge?: bigint;
}

export interface AssurancePolicyConfiguration {
  readonly id: string;
  readonly requirements: readonly AssuranceRequirementConfiguration[];
}

export interface AcceptedRegistryConfiguration {
  readonly principalMethods: readonly string[];
  readonly signatureSuites: readonly string[];
  readonly evidenceTypes: readonly string[];
  readonly principalStatusMethods: readonly string[];
  readonly grantStatusMethods: readonly string[];
  readonly assuranceClaims: readonly string[];
  readonly assuranceImplications?: readonly string[];
  readonly resourceMatchers: readonly string[];
  readonly budgetAlgebras?: readonly string[];
  readonly criticalExtensions?: readonly string[];
  readonly profiles: readonly TrustProfile[];
  readonly profilePolicies: readonly string[];
}

export type VerifierLimitKind =
  | "bundle-bytes"
  | "action-bytes"
  | "context-bytes"
  | "grants"
  | "actions"
  | "plan-leaves"
  | "plan-depth"
  | "plan-branching"
  | "evidence-objects"
  | "evidence-bytes"
  | "control-bindings"
  | "principal-status-statements"
  | "grant-status-statements"
  | "attachments"
  | "attachment-bytes"
  | "signatures"
  | "signature-bytes"
  | "permissions"
  | "audiences"
  | "critical-extensions"
  | "critical-extension-bytes"
  | "allowed-body-digests"
  | "binding-evidence"
  | "canonical-body-bytes"
  | "registry-entries"
  | "trust-anchors";

export interface TrustedContextConfiguration {
  readonly sourceId: string;
  readonly composition: Readonly<{
    minimumAuthorizedBranches: number;
    minimumDistinctActors: number;
    minimumDistinctRoots: number;
  }>;
  readonly trustAnchors: readonly TrustAnchorConfiguration[];
  readonly registries: AcceptedRegistryConfiguration;
  readonly expectedAudience: string;
  readonly evaluationTime: bigint;
  readonly assurance: AssurancePolicyConfiguration;
  readonly principalStatus: PrincipalStatusSnapshot;
  readonly grantStatus: GrantStatusSnapshot;
  readonly resourceMatcher: string;
  readonly profilePolicy: string;
  readonly channelPolicy: string;
  readonly limits?: Readonly<Partial<Record<VerifierLimitKind, number>>>;
  readonly workUnits?: bigint;
}

export interface CompiledTrustedContext {
  readonly source: TrustedContextSource;
  readonly verifierConfiguration: Uint8Array;
  readonly roots: readonly string[];
  readonly offlineBundle: OfflineTrustBundle;
}

const OFFLINE_TRUST_TOKEN = Symbol("auths-offline-trust");
const offlineTrustBytes = new WeakMap<OfflineTrustBundle, Uint8Array>();
let mintOfflineTrustBundle: (
  bytes: Uint8Array,
  provenance: EvidenceProvenance,
) => OfflineTrustBundle;

export interface EvidenceProvenance {
  readonly source: string;
  readonly observedAt: bigint;
  readonly validUntil: bigint;
  readonly version: string;
}

export class OfflineTrustBundle {
  readonly provenance: EvidenceProvenance;

  private constructor(
    token: typeof OFFLINE_TRUST_TOKEN,
    bytes: Uint8Array,
    provenance: EvidenceProvenance,
  ) {
    if (token !== OFFLINE_TRUST_TOKEN) throw new TypeError("sealed Auths trust bundle");
    this.provenance = copyEvidenceProvenance(provenance);
    offlineTrustBytes.set(this, bytes.slice());
    Object.freeze(this);
  }

  private static create(
    token: typeof OFFLINE_TRUST_TOKEN,
    bytes: Uint8Array,
    provenance: EvidenceProvenance,
  ): OfflineTrustBundle {
    return new OfflineTrustBundle(token, bytes, provenance);
  }

  static {
    mintOfflineTrustBundle = (bytes, provenance) =>
      OfflineTrustBundle.create(OFFLINE_TRUST_TOKEN, bytes, provenance);
  }

  export(): Uint8Array {
    const bytes = offlineTrustBytes.get(this);
    if (bytes === undefined) throw new TypeError("invalid Auths trust bundle");
    return bytes.slice();
  }
}

export interface EvidenceSourceRequest {
  readonly sourceId: string;
  readonly signal: AbortSignal;
  readonly maximumBytes: number;
  readonly maximumRedirects: number;
  readonly allowPrivateNetwork: boolean;
}

export interface EvidenceSourceResult {
  readonly bytes: Uint8Array;
  readonly provenance: EvidenceProvenance;
}

export interface EvidenceSourcePort {
  load(request: EvidenceSourceRequest): Promise<EvidenceSourceResult>;
}

export interface EvidenceSourceOptions {
  readonly sourceId: string;
  readonly port: EvidenceSourcePort;
  readonly timeoutMs?: number;
  readonly maximumBytes?: number;
  readonly maximumRedirects?: number;
  readonly allowPrivateNetwork?: boolean;
  readonly signal?: AbortSignal;
  readonly evaluationTime: bigint;
  readonly telemetry?: TelemetryPort;
  readonly correlationId?: string;
}

export async function compileTrustedContext(
  configuration: TrustedContextConfiguration,
): Promise<CompiledTrustedContext> {
  if (configuration === null || typeof configuration !== "object") {
    throw new AuthsWorkflowError("invalid-trusted-context", "trusted context configuration is missing");
  }
  const copied = copyConfiguration(configuration);
  const sourceId = configuration.sourceId;
  const roots = Object.freeze(configuration.trustAnchors.map((anchor) => anchor.principal));
  const principalStatus = statusSnapshotBytes(configuration.principalStatus);
  const grantStatus = statusSnapshotBytes(configuration.grantStatus);
  const engine = await loadPackagedWorkflowEngine();
  let native;
  try {
    native = engine.compileTrustedContextV1(
      copied,
      principalStatus,
      grantStatus,
    );
  } catch {
    throw new AuthsWorkflowError(
      "invalid-trusted-context",
      "Rust rejected the typed trusted context configuration",
    );
  }
  try {
    const context = new Uint8Array(native.cbor);
    const verifierConfiguration = new Uint8Array(native.verifierConfiguration);
    const source = trustedContextSource({
      sourceId,
      provider: Object.freeze({
        async loadTrustedContext(): Promise<Uint8Array> {
          return context.slice();
        },
      }),
    });
    return Object.freeze({
      source,
      verifierConfiguration,
      roots,
      offlineBundle: mintOfflineTrustBundle(context, {
        source: sourceId,
        observedAt: configuration.evaluationTime,
        validUntil: configuration.evaluationTime,
        version: "compiled-v1",
      }),
    });
  } finally {
    native.free?.();
  }
}

/** Acquires bounded evidence through an explicit I/O port and returns inert offline bytes. */
export async function loadOfflineTrustBundle(options: EvidenceSourceOptions): Promise<OfflineTrustBundle> {
  const timeoutMs = options.timeoutMs ?? 5_000;
  const maximumBytes = options.maximumBytes ?? 1_048_576;
  const maximumRedirects = options.maximumRedirects ?? 0;
  if (!Number.isSafeInteger(timeoutMs) || timeoutMs < 1 || timeoutMs > 60_000 ||
      !Number.isSafeInteger(maximumBytes) || maximumBytes < 1 || maximumBytes > 16_777_216 ||
      !Number.isSafeInteger(maximumRedirects) || maximumRedirects < 0 || maximumRedirects > 4 ||
      options.sourceId.length === 0 || options.sourceId.length > 128) {
    throw new AuthsWorkflowError("invalid-trusted-context", "evidence source limits are invalid");
  }
  const controller = new AbortController();
  const correlationId = options.correlationId ?? nextEvidenceCorrelationId();
  const started = performance.now();
  void emitAuthsEvent(options.telemetry, {
    name: "auths.acquisition.started",
    timestamp: Date.now(),
    correlationId,
    operation: "load-trust-evidence",
    stage: "acquisition",
    outcome: "started",
  });
  const abort = () => controller.abort(options.signal?.reason);
  options.signal?.addEventListener("abort", abort, { once: true });
  const timer = setTimeout(() => controller.abort(new DOMException("timed out", "TimeoutError")), timeoutMs);
  try {
    const result = await options.port.load(Object.freeze({
      sourceId: options.sourceId,
      signal: controller.signal,
      maximumBytes,
      maximumRedirects,
      allowPrivateNetwork: options.allowPrivateNetwork ?? false,
    }));
    controller.signal.throwIfAborted();
    if (!(result.bytes instanceof Uint8Array) || result.bytes.length === 0 ||
        result.bytes.length > maximumBytes) {
      throw new AuthsWorkflowError("invalid-trusted-context", "evidence source returned invalid bytes");
    }
    const provenance = copyEvidenceProvenance(result.provenance);
    if (provenance.observedAt > options.evaluationTime || provenance.validUntil < options.evaluationTime) {
      throw new AuthsWorkflowError("invalid-trusted-context", "evidence source returned stale evidence");
    }
    const bundle = mintOfflineTrustBundle(result.bytes, provenance);
    void emitAuthsEvent(options.telemetry, {
      name: "auths.acquisition.completed",
      timestamp: Date.now(),
      correlationId,
      operation: "load-trust-evidence",
      stage: "acquisition",
      outcome: "succeeded",
      durationMs: performance.now() - started,
    });
    return bundle;
  } catch (error) {
    void emitAuthsEvent(options.telemetry, {
      name: "auths.acquisition.failed",
      timestamp: Date.now(),
      correlationId,
      operation: "load-trust-evidence",
      stage: "acquisition",
      outcome: "failed",
      durationMs: performance.now() - started,
    });
    if (error instanceof AuthsWorkflowError) throw error;
    throw new AuthsWorkflowError(
      "trusted-context-source-failed",
      "evidence source operation failed",
      { operation: "load-evidence", stage: "acquisition", retry: "conditional" },
    );
  } finally {
    clearTimeout(timer);
    options.signal?.removeEventListener("abort", abort);
  }
}

let evidenceCorrelationSequence = 0;

function nextEvidenceCorrelationId(): string {
  evidenceCorrelationSequence = (evidenceCorrelationSequence + 1) % Number.MAX_SAFE_INTEGER;
  return `auths-evidence-${Date.now().toString(36)}-${evidenceCorrelationSequence.toString(36)}`;
}

/** Validates imported offline evidence against an exact root and WASM configuration. */
export async function trustedContextFromOfflineBundle(
  sourceId: string,
  bundle: OfflineTrustBundle,
  rootPrincipal: string,
  verifierConfiguration: Uint8Array,
): Promise<TrustedContextSource> {
  const bytes = bundle.export();
  const engine = await loadPackagedWorkflowEngine();
  try {
    engine.validateTrustedContextV1(bytes, rootPrincipal, verifierConfiguration);
  } catch {
    throw new AuthsWorkflowError("invalid-trusted-context", "offline evidence does not match trust inputs");
  }
  return trustedContextSource({
    sourceId,
    provider: Object.freeze({ async loadTrustedContext() { return bytes.slice(); } }),
  });
}

function copyEvidenceProvenance(provenance: EvidenceProvenance): EvidenceProvenance {
  if (provenance.source.length === 0 || provenance.source.length > 512 ||
      provenance.version.length === 0 || provenance.version.length > 128 ||
      provenance.observedAt > provenance.validUntil) {
    throw new AuthsWorkflowError("invalid-trusted-context", "evidence provenance is invalid");
  }
  return Object.freeze({ ...provenance });
}

function copyConfiguration(configuration: TrustedContextConfiguration): object {
  const registries = configuration.registries;
  const limits = Object.entries(configuration.limits ?? {}).map(([kind, value]) => {
    if (!Number.isSafeInteger(value) || (value ?? -1) < 0) {
      throw new AuthsWorkflowError("invalid-trusted-context", "verifier limit is outside bounds");
    }
    return Object.freeze({ kind, value });
  });
  return Object.freeze({
    composition: Object.freeze({
      expectedPlan: undefined,
      ...configuration.composition,
    }),
    trustAnchors: Object.freeze(configuration.trustAnchors.map((anchor) => Object.freeze({
      ...anchor,
      acceptedMethods: Object.freeze([...anchor.acceptedMethods]),
      profiles: copyProfiles(anchor.profiles),
      permissions: Object.freeze(anchor.permissions.map((permission) => Object.freeze({ ...permission }))),
      resourceNamespaces: Object.freeze([...anchor.resourceNamespaces]),
      audiences: Object.freeze([...anchor.audiences]),
      ...(anchor.budget === undefined ? {} : { budget: Object.freeze({ ...anchor.budget }) }),
      statusPolicy: Object.freeze({ ...anchor.statusPolicy }),
    }))),
    registries: Object.freeze({
      principalMethods: Object.freeze([...registries.principalMethods]),
      signatureSuites: Object.freeze([...registries.signatureSuites]),
      evidenceTypes: Object.freeze([...registries.evidenceTypes]),
      principalStatusMethods: Object.freeze([...registries.principalStatusMethods]),
      grantStatusMethods: Object.freeze([...registries.grantStatusMethods]),
      assuranceClaims: Object.freeze([...registries.assuranceClaims]),
      assuranceImplications: Object.freeze([...(registries.assuranceImplications ?? [])]),
      resourceMatchers: Object.freeze([...registries.resourceMatchers]),
      budgetAlgebras: Object.freeze([...(registries.budgetAlgebras ?? [])]),
      criticalExtensions: Object.freeze([...(registries.criticalExtensions ?? [])]),
      profiles: copyProfiles(registries.profiles),
      profilePolicies: Object.freeze([...registries.profilePolicies]),
    }),
    expectedAudience: configuration.expectedAudience,
    evaluationTime: configuration.evaluationTime,
    assurance: Object.freeze({
      id: configuration.assurance.id,
      requirements: Object.freeze(configuration.assurance.requirements.map((requirement) => Object.freeze({ ...requirement }))),
    }),
    resourceMatcher: configuration.resourceMatcher,
    profilePolicy: configuration.profilePolicy,
    channelPolicy: configuration.channelPolicy,
    limits: Object.freeze(limits),
    workUnits: configuration.workUnits ?? 50_000n,
  });
}

function copyProfiles(profiles: readonly TrustProfile[]): readonly Readonly<TrustProfile>[] {
  return Object.freeze(profiles.map((profile) => Object.freeze({ ...profile })));
}

File: bindings/python/python/auths/trust.py
"""Typed trust configuration, evidence, and offline bundles."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Literal, Optional, Protocol, Sequence, Tuple, runtime_checkable

from . import _native as native
from .authority import ProofPlan, _native_proof_plan
from .lifecycle import GrantStatusSnapshot, PrincipalStatusSnapshot
from .workflow import (
    BudgetCeiling,
    Permission,
    Principal,
    Profile,
    TrustedAuthority,
    TrustedAuthoritySnapshot,
)

AssuranceRole = Literal["root", "intermediate", "actor", "external-issuer"]
AssuranceQuantifier = Literal["any", "every"]


@dataclass(frozen=True)
class AssuranceRequirement:
    role: AssuranceRole
    quantifier: AssuranceQuantifier
    claim: str
    maximum_age: Optional[int] = None


@dataclass(frozen=True)
class AssurancePolicy:
    id: str
    requirements: Tuple[AssuranceRequirement, ...]

    def __post_init__(self) -> None:
        object.__setattr__(self, "requirements", tuple(self.requirements))

    def _native(self) -> native.AssurancePolicy:
        return native.AssurancePolicy(
            self.id,
            [
                (value.role, value.quantifier, value.claim, value.maximum_age)
                for value in self.requirements
            ],
        )


@dataclass(frozen=True)
class TrustAnchor:
    id: str
    principal: Principal
    accepted_methods: Tuple[str, ...]
    profiles: Tuple[Profile, ...]
    permissions: Tuple[Permission, ...]
    resource_namespaces: Tuple[str, ...]
    audiences: Tuple[str, ...]
    not_before: int
    expires_at: int
    max_delegation_depth: int
    assurance_policy: str
    budget: Optional[BudgetCeiling] = None
    status: Optional[Tuple[str, int]] = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "accepted_methods", tuple(self.accepted_methods))
        object.__setattr__(self, "profiles", tuple(self.profiles))
        object.__setattr__(self, "permissions", tuple(self.permissions))
        object.__setattr__(self, "resource_namespaces", tuple(self.resource_namespaces))
        object.__setattr__(self, "audiences", tuple(self.audiences))

    def _native(self) -> native.TrustAnchor:
        return native.TrustAnchor(
            self.id,
            self.principal,
            list(self.accepted_methods),
            [(value.id, value.version) for value in self.profiles],
            [(value.capability, value.resource) for value in self.permissions],
            list(self.resource_namespaces),
            list(self.audiences),
            self.not_before,
            self.expires_at,
            None if self.budget is None else (self.budget.algebra, self.budget.value),
            self.max_delegation_depth,
            self.assurance_policy,
            self.status,
        )


@dataclass(frozen=True)
class EvidenceRequest:
    source_id: str
    maximum_bytes: int
    timeout: float
    maximum_redirects: int = 0
    allow_private_network: bool = False


@dataclass(frozen=True)
class EvidenceProvenance:
    source_id: str
    observed_at: int
    valid_until: int
    version: str


@dataclass(frozen=True)
class ResolvedEvidence:
    bytes: bytes
    media_type: str
    provenance: EvidenceProvenance

    def __post_init__(self) -> None:
        object.__setattr__(self, "bytes", bytes(self.bytes))


@runtime_checkable
class EvidenceProvider(Protocol):
    async def resolve(self, request: EvidenceRequest) -> ResolvedEvidence: ...


@dataclass(frozen=True)
class OfflineEvidenceBundle:
    evidence: Tuple[ResolvedEvidence, ...]
    captured_at: int

    def __post_init__(self) -> None:
        values = tuple(self.evidence)
        if not values:
            raise ValueError("offline evidence bundle cannot be empty")
        object.__setattr__(self, "evidence", values)


@dataclass(frozen=True)
class CompiledTrust:
    context: native.TrustedContext
    roots: Tuple[Principal, ...]
    offline_evidence: Optional[OfflineEvidenceBundle]


@dataclass(frozen=True)
class PolicyReplacement:
    current: CompiledTrust
    replacement: CompiledTrust
    activated_at: int


def replace_policy(
    current: CompiledTrust,
    replacement: CompiledTrust,
    *,
    activated_at: int,
) -> PolicyReplacement:
    if type(current) is not CompiledTrust or type(replacement) is not CompiledTrust:
        raise TypeError("policy replacement requires compiled trust values")
    if activated_at < 0:
        raise ValueError("policy activation time cannot be negative")
    if bytes(native.inspect_trusted_context(current.context)) == bytes(
        native.inspect_trusted_context(replacement.context)
    ):
        raise ValueError("replacement policy must have a different configuration")
    return PolicyReplacement(current, replacement, activated_at)


async def load_evidence(
    provider: EvidenceProvider, request: EvidenceRequest
) -> ResolvedEvidence:
    if (
        request.maximum_bytes < 1
        or request.maximum_bytes > 16 * 1024 * 1024
        or request.timeout <= 0
        or request.timeout > 300
        or request.maximum_redirects < 0
        or request.maximum_redirects > 8
    ):
        raise ValueError("evidence request is outside supported bounds")
    result = await asyncio.wait_for(provider.resolve(request), request.timeout)
    if type(result) is not ResolvedEvidence:
        raise TypeError("evidence provider returned the wrong type")
    if (
        len(result.bytes) > request.maximum_bytes
        or result.provenance.source_id != request.source_id
        or result.provenance.valid_until < result.provenance.observed_at
    ):
        raise ValueError("evidence provider returned inconsistent evidence")
    return result


def compile_trust(
    *,
    anchors: Sequence[TrustAnchor],
    assurance: AssurancePolicy,
    minimum_authorized_branches: int = 1,
    minimum_distinct_actors: int = 1,
    minimum_distinct_roots: int = 1,
    expected_plan: Optional[ProofPlan] = None,
    principal_status: Optional[PrincipalStatusSnapshot] = None,
    grant_status: Optional[GrantStatusSnapshot] = None,
    channel_policy: str = "none-v1",
    evidence_types: Sequence[str] = (),
    critical_extensions: Sequence[str] = (),
    offline_evidence: Optional[OfflineEvidenceBundle] = None,
) -> CompiledTrust:
    anchor_values = tuple(anchors)
    if not anchor_values or any(
        type(value) is not TrustAnchor for value in anchor_values
    ):
        raise ValueError("trust requires at least one typed anchor")
    context = native.compile_trusted_context(
        native.self_contained_configuration(),
        None if expected_plan is None else _native_proof_plan(expected_plan),
        minimum_authorized_branches,
        minimum_distinct_actors,
        minimum_distinct_roots,
        [value._native() for value in anchor_values],
        assurance._native(),
        None if principal_status is None else principal_status._native,
        None if grant_status is None else grant_status._native,
        channel_policy,
        list(evidence_types),
        list(critical_extensions),
    )
    return CompiledTrust(
        context,
        tuple(value.principal for value in anchor_values),
        offline_evidence,
    )


StatusSnapshot = native.StatusSnapshot
TrustedContext = native.TrustedContext
compile_trusted_context = native.compile_trusted_context
parse_trusted_context = native.parse_trusted_context
self_contained_configuration = native.self_contained_configuration
status_snapshot = native.status_snapshot

__all__ = [
    "AssurancePolicy",
    "AssuranceQuantifier",
    "AssuranceRequirement",
    "AssuranceRole",
    "CompiledTrust",
    "EvidenceProvenance",
    "EvidenceProvider",
    "EvidenceRequest",
    "OfflineEvidenceBundle",
    "PolicyReplacement",
    "ResolvedEvidence",
    "StatusSnapshot",
    "TrustAnchor",
    "TrustedAuthority",
    "TrustedAuthoritySnapshot",
    "TrustedContext",
    "compile_trust",
    "compile_trusted_context",
    "load_evidence",
    "parse_trusted_context",
    "replace_policy",
    "self_contained_configuration",
    "status_snapshot",
]
